### PR TITLE
Tutorial 2.0

### LINKED ISSUE #253: Tutorial JSON Loading
design tutorial JSON and load them in

### PR DESCRIPTION
**Pull request by ssorenson130**
### Summary 
Implement the new tutorial levels specified in the google doc.
Close #253 

### Testing 
Tested with 2 and 3 players on windows
### Review Focus 

Please walk through all tutorial levels once if you can. It's all hardcoded so the code is mildly disgusting.

### CODE REVIEW

File: source/GLaDOS.cpp [hunk 1]
```diff
@@ -276,9 +392,112 @@ void GLaDOS::update(float dt) {
...
+						ship->getSize() / 2 - abs(abs(suggestedAngle - mid) - ship->getSize() / 2);
+					if (diff > width / 2) {
+						// clamp this angle within the width of the section
+						suggestedAngle -= width;
+						if (suggestedAngle < 0) suggestedAngle += ship->getSize();
+					}
+					placeObject({BuildingBlockModel::Breach, 0, -1}, suggestedAngle,
+								(i + 1) % ship->getDonuts().size());
+				}
+				things--;
```
Comment (mt-xing) [inline source/GLaDOS.cpp:419]: What is `things`? Could it get a more descriptive name?

File: source/GameMode.cpp [hunk 2]
```diff
@@ -233,6 +241,8 @@ void GameMode::update(float timestep) {
 		if (!donutModel->isJumping() && playerID != breach->getPlayer() &&
 			diff < globals::BREACH_WIDTH && breach->getHealth() != 0) {
 			// Slow player by drag factor
+			CULog("Player %d at angle %f slowed by breach at %f with id %d", playerID,
+				  donutModel->getAngle(), breach->getAngle(), breach->getPlayer());
```
Comment (mt-xing) [inline source/GameMode.cpp:245]: This seems like a potentially high frequency log. Do we want that in here?

File: source/MagicInternetBox.h [hunk 3]
```diff
@@ -406,6 +408,16 @@ class MagicInternetBox {
...
 
+	/**
+	 * Inform the host that a task requiring all members of the ship has succeeded
+	 */
+	void succeedAllTask();
+
+	/**
+	 * Inform the clients that the level has been won
+	 */
+	void forceWinLevel();
```
Comment (mt-xing) [inline source/MagicInternetBox.h:419]: Why is this necessary? Shouldn't each client have its own version of the state?

File: source/UnopenableNode.cpp [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+	addChild(p);
+	p->setAnchor(Vec2::ANCHOR_BOTTOM_CENTER);
+	p->setPosition(0, 0);
+	setScale(DOOR_SCALE);
+	return true;
+}
+
+void UnopenableNode::prePosition() { angle = unopModel->getAngle(); }
+
+void UnopenableNode::postPosition() {}
```
Comment (mt-xing) [inline source/UnopenableNode.cpp:29]: Don't override a method you're not planning on using.

File: source/MagicInternetBox.h (hunk 3 above)
Comment (ssorenson130) [reply source/MagicInternetBox.h:419]: Yes, but since we have tutorial levels that rely on events being completed (not time) and the host generates those events, the host needs to tell the clients when all is done.

File: source/GameMode.cpp (hunk 2 above)
Comment (ssorenson130) [reply source/GameMode.cpp:245]: ah, my bad. should've been deleted

File: source/GLaDOS.cpp (hunk 1 above)
Comment (ssorenson130) [reply source/GLaDOS.cpp:419]: well it's described in the associated documentation/comment, but I'll change it just for you

**Comment by ssorenson130** [comment]
> This does not compile. There's a variable in `GLaDOS.cpp` called `customEventCtr` that is not defined anywhere. Confirmed on both Windows and macOS.

my bad, forgot to add a file to git

**Comment by mt-xing** [comment]
image
The first tutorial level probably shouldn't spawn the two breaches too close to each other.

**Comment by mt-xing** [comment]
Also my client just crashed entering the button tutorial level.
